libraries/AP_UAVCAN/AP_UAVCAN_Servers.cpp

```diff
@@ -0,0 +1,309 @@
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Siddharth Bharat Purohit
+ */
+
+#include <AP_HAL/AP_HAL.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "AP_UAVCAN_Servers.h"
+
+#ifdef HAS_UAVCAN_SERVERS
+
+#include <uavcan/protocol/dynamic_node_id_server/event.hpp>
+#include <uavcan/protocol/dynamic_node_id_server/storage_backend.hpp>
+#include <uavcan/protocol/dynamic_node_id_server/centralized.hpp>
+#include <uavcan/protocol/HardwareVersion.hpp>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+#include <AP_BoardConfig/AP_BoardConfig_CAN.h>
+
+#include <AP_Common/AP_Common.h>
+#include <DataFlash/DataFlash.h>
+
+#if HAL_OS_POSIX_IO
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <dirent.h>
+#if defined(__APPLE__) && defined(__MACH__)
+#include <sys/param.h>
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#endif
+#endif
+
+#if HAL_OS_FATFS_IO
+#include <stdio.h>
+#endif
+
+#ifndef UAVCAN_NODE_DB_PATH
+#define UAVCAN_NODE_DB_PATH HAL_BOARD_STORAGE_DIRECTORY "/UAVCAN"
+#endif
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_uavcan(fmt, args...) do { hal.console->printf(fmt, ##args); } while (0)
+
+class AP_UAVCAN_CentralizedServer : public uavcan::dynamic_node_id_server::CentralizedServer
+{
+public:
+    AP_UAVCAN_CentralizedServer(uavcan::INode& node, uavcan::dynamic_node_id_server::IStorageBackend &storage_backend, uavcan::dynamic_node_id_server::IEventTracer &tracer) :
+        uavcan::dynamic_node_id_server::CentralizedServer(node, storage_backend, tracer) {}
+};
+
+class AP_UAVCAN_FileEventTracer : public uavcan::dynamic_node_id_server::IEventTracer
+{
+protected:
+    virtual void onEvent(uavcan::dynamic_node_id_server::TraceCode code, uavcan::int64_t argument)
+    {
+        DataFlash_Class::instance()->Log_Write("UCEV", "TimeUS,code,arg", "s--", "F--", "Qhq", AP_HAL::micros64(), code, argument);
+    }
+};
+
+class AP_UAVCAN_FileStorageBackend : public uavcan::dynamic_node_id_server::IStorageBackend
+{
+    /**
+     * Maximum length of full path including / and key max
+     */
+    enum { MaxPathLength = 128 };
+
+    enum { FilePermissions = 438 };     ///< 0o666
+
+    enum { MaxNumOpens = 100 };
+    /**
+     * This type is used for the path
+     */
+    typedef uavcan::MakeString<MaxPathLength>::Type PathString;
+
+    PathString base_path;
+
+    static uint8_t num_opens;
+protected:
+    virtual String get(const String& key) const
+    {
+        using namespace std;
+        PathString path = base_path.c_str();
+        path += key;
+        String value;
+        //This is to deter frequent inflight opening and closing of files during an event
+        //where the device is misbehaving
+        num_opens++;
```

Comment:
as this is 8 bit it will start doing opens again when it wraps. Need to increment after the check

Reply:
yes!